### ISSUE TITLE
Support mirroring across clusters with overlapping networks

### ISSUE THREAD

**Issue by travisn** [rook/rook#11070]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Kubernetes clusters are frequently setup with a standard networking configuration, which results in clusters having overlapping networks. For example, the subnet of one cluster may be `10.0.0.0/16`. Internally to the cluster network there will be no conflicts, but another cluster may also be using `10.0.0.0/16`. If network config attempts to connect them together, for example with Multus or Submariner, the connections will work fine as long as a connection only needs to be addressable either from outside or inside one of the networks. However, there is not an IP that will work both external and internal to a cluster. 

The Ceph mons require an IP address as part of their identity. If the mons are going to be addressable from external to the cluster, the mons must recognize the IP address that was used by the client to connect to the mon, otherwise the connections will be rejected. The network tunnel between clusters will make it impossible to use the same IP from the external cluster as internal to the cluster for the local mon access.

This is not an issue when using host networking in a traditional Ceph configuration since all mons would be addressable by their host IP address. 

A potential solution discussed with the RADOS team is to add a second entry in the monmap. There would be one entry for the mon IP in the local cluster, and another IP added to the monmap for the IP addressable from the external cluster. All requests coming from the external cluster would respond on that external IP, while internal cluster requests would respond on the local cluster IP. This is similar to the approach taken for messenger v1 and v2, where each one had an entry per mon.

@gregsfortytwo @jdurgin @neha-ojha @mmgaggle Any other ideas since our discussion, or clarifications on the proposal? 

We can open a Ceph tracker for this, but wanted to start the discussion in Rook since it's a common K8s issue for enabling mirroring between clusters.

**What is use case behind this feature:**
Enable mirroring between K8s clusters where the networks are overlapping.

**Comment by ideepika** [comment]
for some more context, rbd mirroring actually uses librados C++ api to configure: fsid, client id, mon host and key in form of token, which is then consumed on secondary cluster https://github.com/ceph/ceph/blob/main/src/librbd/api/Mirror.cc#L1289-L1514

**Comment by idryomov** [comment]
> The Ceph mons require an IP address as part of their identity. If the mons are going to be addressable from external to the cluster, the mons must recognize the IP address that was used by the client to connect to the mon, otherwise the connections will be rejected. The network tunnel between clusters will make it impossible to use the same IP from the external cluster as internal to the cluster for the local mon access.

I don't think this is specific to the monitors.  The "must recognize the IP address that was used by the client to connect" restriction is coming from Ceph messenger (this is an excerpt from `ProtocolV2.cc` but IIRC `ProtocolV1.cc` has a similar check):

```
  if (!messenger->get_myaddrs().contains(client_ident.target_addr())) {
    ldout(cct,5) << __func__ << " peer is trying to reach "
		 << client_ident.target_addr()
		 << " which is not us (" << messenger->get_myaddrs() << ")"
		 << dendl;
    return _fault();
  }
```

So I believe this goes for any Ceph daemon that clients can connect to, most notably OSDs.

> 
> This is not an issue when using host networking in a traditional Ceph configuration since all mons would be addressable by their host IP address.
> 
> A potential solution discussed with the RADOS team is to add a second entry in the monmap. There would be one entry for the mon IP in the local cluster, and another IP added to the monmap for the IP addressable from the external cluster. All requests coming from the external cluster would respond on that external IP, while internal cluster requests would respond on the local cluster IP. This is similar to the approach taken for messenger v1 and v2, where each one had an entry per mon.

Putting aside the need to repeat this exercise for at least osdmap (and likely mdsmap and mgrmap as well), I'm not clear how this approach handles:

- bidirectional mirroring, where mirroring daemons on cluster A must be able to talk to all monitors and OSDs on cluster B and mirroring daemons on cluster B must be able to talk to all monitors and OSDs on cluster A at the same time.  Does the same pair of addresses work in both directions?
- more than one external cluster, where images on cluster A are mirrored to both cluster B and cluster C.  Is the IP addressable from the external cluster specific to that cluster (i.e. that tunnel)?

**Comment by gregsfortytwo** [comment]
> > The Ceph mons require an IP address as part of their identity. If the mons are going to be addressable from external to the cluster, the mons must recognize the IP address that was used by the client to connect to the mon, otherwise the connections will be rejected. The network tunnel between clusters will make it impossible to use the same IP from the external cluster as internal to the cluster for the local mon access.
> 
> I don't think this is specific to the monitors. The "must recognize the IP address that was used by the client to connect" restriction is coming from Ceph messenger (this is an excerpt from `ProtocolV2.cc` but IIRC `ProtocolV1.cc` has a similar check):
> 
> ```
>   if (!messenger->get_myaddrs().contains(client_ident.target_addr())) {
>     ldout(cct,5) << __func__ << " peer is trying to reach "
> 		 << client_ident.target_addr()
> 		 << " which is not us (" << messenger->get_myaddrs() << ")"
> 		 << dendl;
>     return _fault();
>   }
> ```
> 
> So I believe this goes for any Ceph daemon that clients can connect to, most notably OSDs.

Yeah. I really don't like the idea, but the cheapest solution in terms of developer effort we came up with was to add yet another messenger to all the daemons, that could be given the "external IP". And hopefully submariner could set up that IP for all external clusters.

This has some notable costs, of course:
* inflating all the maps more, for the extra IP addresses
* adding another messenger/dispatcher to all the daemons, which has memory overheads (though probably not a lot of CPU costs, since it hasn't caused problems in the OSD cluster/public pair)

The monitors are special in that their IP is tied to their identity, whereas the OSDs are perfectly happy to use whatever IP addresses they get told they look like. So this would require extra work so that the monitors can talk on extra addresses without complaining about the mismatch.

Also, any of these solutions will either require updated clients (to read the new fields), or else for the monitors to generate "fake" maps for the external clients which embed the external IPs in place of the internal ones. (I'm not sure if embedding the "fake" IPs is actually possible due to the checksums we embed in the maps, and the client/OSD sharing.)

**Comment by travisn** [comment]
Addressing this scenario is obviously going to take some time to design and implement. Is there any workaround that could be considered in the meantime? For example, could the check for mon identity be relaxed so it could accept rerouted requests to the mons? Or perhaps that doesn't get us very far since without handling it in the monmap, the mons just really wouldn't know how to route the clients to the osds or other daemons across cluster?

**Comment by BlaineEXE** [comment]
FWIW, this concern is not specifically limited to the monitors, though they are certainly the most important component. We should consider any component that may need to be addressed from both inside and outside of a Kubernetes cluster. Specifically, I think we need to consider the CephFS MDSes. We might consider the MGR as well, though I'm less familiar with this. I don't believe we need to consider RGW; we seem to already be able to do internal and external routing successfully for those.

**Comment by travisn** [comment]
From the discussion earlier today, my takeaway is that there are two overall approaches to consider. 

### Fully Addressable

Make all Ceph daemons addressable across clusters. Every mon, osd, and mgr daemon would require an endpoint that is globally addressable. If we are interested in cross-cluster scenarios other than mirroring, those daemons would also need to be addressable. Possible implementations could be to update the msgr as discussed earlier in this thread where it will be a very involved update that and take a long time to implement and stabilize. 

This fully connected approach does not seem to follow current submariner designs as there seem to be restrictions on the tunneling through a single daemon. We would need more scalability from submariner to handle the distributed nature of the connections.

### RBD Mirror Gateway

Configure the rbd mirror daemon to talk to a gateway in the remote cluster. Only the rbd mirror daemon would communicate between clusters. RADOS would not be affected since mons and osds would not be exposed between clusters. Of course, this will limit the scalability of the mirroring since there is no advantage of the distributed network with ceph daemons. This approach also follows patterns already established by submariner where it is simple to configure a single gateway.

**Comment by idryomov** [comment]
> ### RBD Mirror Gateway
> 
> Configure the rbd mirror daemon to talk to a gateway in the remote cluster. Only the rbd mirror daemon would communicate between clusters. RADOS would not be affected since mons and osds would not be exposed between clusters.

It's important to get the terminology right here.  The way I see it, this would be a RADOS gateway, not an RBD Mirror gateway.  This new gateway must be optional, so rbd-mirror daemon should continue to speak RADOS on the backend as it does today.  The new gateway should therefore be a generic RADOS gateway, essentially combining all monitor, manager, OSD, MDS, etc endpoints in a single process that can be set up the way Submariner expects.  Any Ceph client (rbd-mirror daemon, cephfs-mirror daemon, CLI commands, etc) should be able to be configured to talk to the new gateway instead of the actual cluster with a one-line change to `ceph.conf` -- substituting the actual monitor addresses with the new gateway address in `mon host`.

**Comment by gregsfortytwo** [comment]
> > ### RBD Mirror Gateway
> > Configure the rbd mirror daemon to talk to a gateway in the remote cluster. Only the rbd mirror daemon would communicate between clusters. RADOS would not be affected since mons and osds would not be exposed between clusters.
> 
> It's important to get the terminology right here. The way I see it, this would be a RADOS gateway, not an RBD Mirror gateway.

While I'd love to not have to develop a separate CephFS gateway, this wasn't the impression I had from that discussion — I thought we were talking about the rbd mirror daemon growing some kind of protocol (or, more probably, a wrapper interface somewhere in its code base?) to handle this.

Happily, in further discussions it sounds like we can get the full networking we need with GlobalNet and setting up OSDs as services (similarly to what Rook already does with monitors), so we currently don't plan to need any proxies. :+1:

**Comment by travisn** [comment]
Fixed by #11578. See the design [here](https://github.com/rook/rook/blob/master/design/ceph/multi-cluster-service.md).
